tests/ctst/common/common.ts

```diff
@@ -31,25 +31,33 @@ export async function cleanS3Bucket(
     if (!bucketName) {
         return;
     }
-    world.resetCommand();
-    world.addCommandParameter({ bucket: bucketName });
-    const createdObjects = world.getSaved<Map<string, string>>('createdObjects');
-    if (createdObjects !== undefined) {
-        const results = await S3.listObjectVersions(world.getCommandParameters());
-        const res = safeJsonParse<ListObjectVersionsOutput>(results.stdout);
-        assert(res.ok);
-        const versions = res.result!.Versions || [];
-        const deleteMarkers = res.result!.DeleteMarkers || [];
-        await Promise.all(versions.concat(deleteMarkers).map(obj => {
-            world.addCommandParameter({ key: obj.Key });
-            world.addCommandParameter({ versionId: obj.VersionId });
-            return S3.deleteObject(world.getCommandParameters());
-        }));
-        world.deleteKeyFromCommand('key');
-        world.deleteKeyFromCommand('versionId');
+    try {
+        Identity.useIdentity(IdentityEnum.ACCOUNT, world.getSaved<string>('accountName') ||
+            world.parameters.AccountName);
+        world.resetCommand();
+        world.addCommandParameter({ bucket: bucketName });
+        const createdObjects = world.getCreatedObjects();
+        if (createdObjects !== undefined) {
+            const results = await S3.listObjectVersions(world.getCommandParameters());
+            const res = safeJsonParse<ListObjectVersionsOutput>(results.stdout);
+            if (!res.ok) {
+                throw results;
+            }
+            const versions = res.result!.Versions || [];
+            const deleteMarkers = res.result!.DeleteMarkers || [];
+            await Promise.all(versions.concat(deleteMarkers).map(obj => {
+                world.addCommandParameter({ key: obj.Key });
+                world.addCommandParameter({ versionId: obj.VersionId });
+                return S3.deleteObject(world.getCommandParameters());
+            }));
+            world.deleteKeyFromCommand('key');
+            world.deleteKeyFromCommand('versionId');
+        }
+        await S3.deleteBucketLifecycle(world.getCommandParameters());
+        await S3.deleteBucket(world.getCommandParameters());
+    } catch (err) {
+        world.logger.warn('Error cleaning bucket', { bucketName, err });
```

Comment:
whu could this happen?
except a bug in the code, it should not happen IMHO.... so having an error is actually a good thing!

Even if it may not be related to the "current" test, this should not be masked (let's be honest: we will never notice a warning, unless we have another issue)... Esp. if it allows to "detect" unexpected system instability (like high cpu usage or pod restarts), it is better to have an error here and fix the root cause of the instability.

Reply:
Some tests actually test the deletion of a bucket, or delete a bucket in their logic; in this case and because this can be dynamic, we do not want to fail a successful scenario in this case, because we get a "NoSuchBucket" error. I don't want to filter out this error as well because it may be a real bug in some other scenarios. So having a warning log seem the good trade-off.

Reply:
> I don't want to filter out this error as well because it may be a real bug in some other scenarios. So having a warning log seem the good trade-off.

Correct me if I am wrong, but here it seems to me we are filtering all errors : so we will filter-out `NoSuchBucket` errors as well as all the others, and just adding a log (which we will not look at unless we have another test failure)
→ If there is a bug causing NoSuchBucket we will not _notice_
→ If there is a bug causing other error, we will not _notice_ either

> we do not want to fail a successful scenario in this case, because we get a "NoSuchBucket" error

If this was the test, it would fail systematically ; yet the problem we are facing now is flaky: so I don't understand how it relates..

> Some tests actually test the deletion of a bucket, or delete a bucket in their logic

If one test deletes a bucket (or object), it should delete it from the "world" ; or we may add a tag/label on the test, to skip the cleanup when that tag is present...

Reply:
> Correct me if I am wrong, but here it seems to me we are filtering all errors : so we will filter-out NoSuchBucket errors as well as all the others, and just adding a log (which we will not look at unless we have another test failure)
→ If there is a bug causing NoSuchBucket we will not notice
→ If there is a bug causing other error, we will not notice either

Exactly, I do not want any scenario to fail due to unrelated logic during cleanup: we already have scenarios covering the DeleteBucket API. Throwing errors will not bring anything to our flakiness nor help us detecting bugs for this API, as it should be covered by dedicated scenarios. We can also have buckets where compliance locking is enabled and so we cannot remove the objects. It is totally fine to keep some buckets when we cannot naively clean them.

> If this was the test, it would fail systematically ; yet the problem we are facing now is flaky: so I don't understand how it relates..

I think your point is not about flakiness (it is already solved) nor bug detection (we have dedicated scenarios - the cleanup will fail if these scenarios fail as well), but about enforcing that each scenario complies with the cleanup logic; i.e, we do not leave the "saved" variables in a state where we still have a reference to a bucket that was deleted. But we do not want to delete the bucket from the world after we delete it, because we might need it again (e.g., bucket re-creation). We can add a new flag to say "this bucket was deleted", but it imposes a new logic in our tests (logic that is unknown unless we actually run the scenario and end up with a failure because we tried to delete an already-deleted bucket).
So I think it is better not to alter any existing scenario, and be very flexible. We rely on the dedicated scenarios to detect bugs around bucket cleaning.

Reply:
>  fail due to unrelated logic during cleanup: we already have scenarios covering the DeleteBucket API

when I talk about bug, it is not necessarily a bug in the DeleteBucket API (which should be covered explicitely indeed), but more about other bugs... e.g. if the system is unstable or the test itself has an issue, being strict allows to surface these issues.

the point is that we cannot garantee that the tests scenarios will cover any bug : so unless there is a known reason to fail, it should not fail... and it is better to fail (even if hard to detect) than ignore a possible bug (which by definition is not covered by our tests). and leaving the bucket unexpectedly present may still cause more errors later, as we would not be in the "nominal" case of the tests.

> We can add a new flag to say "this bucket was deleted", but it imposes a new logic in our tests (logic that is unknown to the dev unless we actually run the scenario and end up with a failure because we tried to delete an already-deleted bucket)

it is unknown until we tell them it is how it works; and if they don't, they will soon discover it since the test will fail.
the real question here is how many test (or test suites?) would be impacted: do we really have many scenarios which delete their own bucket?

(also I guess deleting a bucket is done is only a few dedicated steps: si it should be handled there, and further limit the changes...)

Reply:
I am doing it, but:

- I am not aligned for the reasons above, plus ->
  - If the test itself has an issue, it should fail; if not the test should be re-implemented. We shouldn't rely on hooks to detect that.
  - At first, buckets were not cleaned for a conscious reason: we want to have a "living platform" where multiple operations are affecting multiple resources, to be realistic. That's why each scenario today is fully independent from others (thanks to random bucket names, or different accounts when required), excluding lifecycle. If we have a bucket that is failing to be deleted, it does not mean that the scenario is responsible for this error. So we might spend more time trying to understand links between a failure and some transient state of the system, than letting the right scenarios fail. Sometimes we could have an overload in the CI (or anything, due to very valid reasons) and thus make this cleanup fail. In this case, the goal is to still pass the other scenarios, and not fail the current one. The philosophy behind CTST is that we are testing what customers do: instabilities happen, other operations might exhaust the resources... That's fine: as long as the scenario achieves its objectives, then the platform is robust enough. CTST tests feature scenarios, what the customer will use. We should only fail a scenario when its objectives are not achieved. If the instability has no effect on the scenario results, all good. If we want to remove all instabilities sources, CTST is not the right tool...

Also we should really unify our test suites across branches, and not over-engineer our scenarios we have in 2.7+ branches: they re-implement lots of existing logic, do not follow best practices of being independent and parallelizable (such as lifecycle - we already know that), making it very slow to integrate changes targeting all branches, due to both conflicts and errors during test execution, as it is hard to track dependencies between a step and what it "requires" in order to work. E.g. here, when we modify the "saved" resource, we end up breaking tests that re-implemented the logic in a different place...

Reply:
Done with added commits specific for 2.7+ branch: https://github.com/scality/zenko/compare/development/2.7...w/2.7/bugfix/ZENKO-4898